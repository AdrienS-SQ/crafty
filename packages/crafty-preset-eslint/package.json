{
    "name": "@swissquote/crafty-preset-eslint",
    "version": "1.6.0",
    "license": "Apache-2.0",
    "repository": "https://github.com/swissquote/crafty.git",
    "publishConfig": {
      "access": "public"
    },
    "main": "src/index.js",
    "dependencies": {
      "@swissquote/crafty-preset-prettier": "1.6.0",
      "@swissquote/eslint-plugin-swissquote": "1.6.0",
      "debug": "^4.1.0",
      "eslint": "5.16.0",
      "eslint-loader": "2.1.2",
      "rollup-plugin-eslint": "5.1.0",
      "resolve-from": "5.0.0",
      "merge": "1.2.1",
      "tmp": "0.1.0"
    },
    "engines": {
      "node": ">=8.0.0"
    }
  }
