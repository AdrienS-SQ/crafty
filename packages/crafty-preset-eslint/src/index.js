const merge = require("merge");
const tmp = require("tmp");
const fs = require("fs");
const path = require("path");

const MODULES = path.join(__dirname, "..", "node_modules");

module.exports = {
  defaultConfig(config) {
    // For some reason, eslint.CLIEngine doesn't support "extends",
    // it has to be done through a configuration file
    const esConfig = tmp.fileSync({
      prefix: "crafty-config-",
      postfix: ".json"
    }).name;

    fs.writeFileSync(
      esConfig,
      JSON.stringify({ extends: ["plugin:@swissquote/swissquote/recommended"] })
    );

    return {
      // In case of browser support changes, we prefer having a fixed version
      eslintReactVersion: "15.0",

      // ESLint Override Rules
      eslint: {
        useEslintrc: false,
        plugins: ["@swissquote/swissquote"],
        configFile: esConfig
      }
    }
  },
  config(config) {
    // Add eslint react version
    config.eslint = merge.recursive(config.eslint, {
      baseConfig: {
        settings: { react: { version: config.eslintReactVersion } }
      }
    });

    return config;
  },
  ide() {
    const configurationBuilder = require("./eslintConfigurator");

    return {
      ".eslintrc.js": {
        content: configurationBuilder(process.argv).configuration,
        serializer: content => `// This configuration was generated by Crafty
// This file is generated to improve IDE Integration
// You don't need to commit this file, nor need it to run \`crafty build\`

module.exports = ${JSON.stringify(content, null, 4)};
`
      }
    };
  },
  commands() {
    return {
      jsLint: {
        //eslint-disable-next-line no-unused-vars
        command: function(crafty, input, cli) {
          global.craftyConfig = crafty.config;
          require("./commands/jsLint");
        },
        description: "Lint JavaScript for errors"
      }
    };
  },
  rollup(crafty, bundle, rollupConfig) {
    // Add Eslint configuration
    //TODO :: throw error after all files are linted, not after first error
    rollupConfig.input.plugins.eslint = {
      plugin: require("rollup-plugin-eslint").eslint,
      weight: -20,
      options: Object.assign({}, crafty.config.eslint, {
        throwOnError: crafty.getEnvironment() === "production",
        exclude: ["node_modules/**"],
        include: ["**/*.js", "**/*.jsx"]
      })
    };
  },
  webpack(crafty, bundle, chain) {
    chain.resolve.extensions.add(".jsx");
    chain.resolve.modules.add(MODULES);
    chain.resolveLoader.modules.add(MODULES);

    // JavaScript linting
    chain.module
      .rule("lint-js")
      .pre()
      .test(/\.jsx?$/)
      .exclude.add(/(node_modules|bower_components)/)
      .end()
      .use("eslint")
      .loader(require.resolve("eslint-loader"))
      .options(crafty.config.eslint);
  }
}