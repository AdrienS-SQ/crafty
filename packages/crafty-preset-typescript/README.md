<table>
<tr><th>Compatible Runners</th><td>

- [rollup.js](05_Packages/02_crafty-runner-rollup.md)
- [Webpack](05_Packages/02_crafty-runner-webpack.md)

</td></tr>
<tr><th>Test Runners</th><td>

- [Jest](05_Packages/05_crafty-preset-jest.md)

</td></tr>
<tr><th>Linters</th><td>

Provides TSLint, with a minimal configuration (should be extended in the future to match the one in stylelint)

</td></tr>
<tr><th>Commands</th><td>

- `tsLint`: Lint TypeScript files, this is a facade for TSLint, pre-configured
  with our preset.

</td></tr>
</table>

[TOC]

## Description

Microsoft presents TypeScript as "JavaScript that scales".
TypeScript is a superset of JavaScript that allows you to type your code.
Not everything needs to have a type as the tool is powerful enough to infer most of the types.

TypeScript has types but it also supports the full EcmaScript 2015 Specification.

## Features

TypeScript being a superset of JavaScript, features supported by [our Babel preset](../05_crafty-preset-babel/JavaScript_Features.md) apply here.

Or you can check out the [TypeScript specific features](TypeScript_Features.md)

## Linting

Like any language, TypeScript has best practices and practices that you shouldn't do.
With the help of `tslint` we check the code for common mistakes and formatting errors.

[Read more](TypeScript_Linting.md)

We don't provide a direct access to TSLint's CLI but plan to do so in the future.

## Installation

```bash
npm install @swissquote/crafty-preset-typescript --save
```

```javascript
module.exports = {
  presets: [
    "@swissquote/crafty-preset-typescript",
    "@swissquote/crafty-runner-webpack", // optional
    "@swissquote/crafty-runner-rollup" // optional
  ],
  js: {
    app: {
      runner: "webpack", // Webpack or rollup.js (optional if you have a single runner defined)
      source: "js/app.ts"
    }
  }
};
```

## Usage with Webpack / rollup.js

Both offer the same level of integration.

You have to configure TypeScript with the `tsconfig.json` file that you put at the root of the project.

Some options can't be auto-configured through the loaders / plugins so that's why you have to do it yourself.
Moreover, the integration with your IDE is simplified if you have a complete `tsconfig.json` in your project.

## Usage with Jest

If you include both `crafty-preset-typescript` and `crafty-preset-jest`.
When running your tests with `crafty test` this preset will be use to convert all `.ts` and `.tsx` files (source and test files)

## Type definition files

TypeScript is also able to understand Types on libraries written in pure JavaScript this is done through types definition files.

Some packages come with their own definitions, for the others there are two ways to consume these files.

1.  Use `@types` scoped packages from NPM
1.  Create your own `*.d.ts` file

[Read more](TypeScript_Typings.md)

## Getting Started

We have a user guide to get started with TypeScript in your project

[Follow the guide](Getting_Started_with_TypeScript.md)

### `crafty tsLint`

This linter will leverage TSLint to lint your TypeScript files with the
Swissquote presets pre-configured. All
[TSLint CLI](https://palantir.github.io/tslint/usage/cli/) options are valid
here.

The additions made by this command are pre-configured rules from the Swissquote
Guidelines.

```bash
crafty tsLint src/**
```

If you pass the `--fix` flag it will fix all the errors it can and write them
directly to the source file.
