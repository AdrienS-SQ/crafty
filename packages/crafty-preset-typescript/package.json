{
  "name": "@swissquote/crafty-preset-typescript",
  "version": "1.3.0",
  "license": "Apache-2.0",
  "repository": "https://github.com/swissquote/crafty.git",
  "publishConfig": {
    "access": "public"
  },
  "main": "src/index.js",
  "dependencies": {
    "@babel/core": "7.0.1",
    "@babel/runtime": "7.0.0",
    "@swissquote/babel-preset-swissquote": "1.3.0",
    "babel-loader": "8.0.2",
    "gulp-concat": "2.6.1",
    "gulp-newer": "1.4.0",
    "gulp-sourcemaps": "2.6.4",
    "gulp-tslint": "8.1.3",
    "gulp-typescript": "5.0.0-alpha.3",
    "rollup-plugin-tslint": "^0.1.24",
    "rollup-plugin-typescript2": "0.17.0",
    "tmp": "0.0.33",
    "ts-jest": "23.1.4",
    "ts-loader": "5.1.0",
    "tslint": "5.11.0",
    "tslint-eslint-rules": "5.4.0",
    "tslint-loader": "3.6.0",
    "typescript": "3.0.3"
  },
  "engines": {
    "node": ">=8.0.0"
  }
}
