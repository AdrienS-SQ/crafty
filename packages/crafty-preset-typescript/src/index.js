const path = require("path");

const createTask = require("./gulp");
const createTempFile = require("./utils").createTempFile;

const MODULES = path.join(__dirname, "..", "node_modules");

function resolve(relative) {
  return path.resolve(process.cwd(), relative);
}

function absolutePath(item) {
  return path.isAbsolute(item) ? item : path.join(process.cwd(), item);
}

module.exports = {
  defaultConfig() {
    return {
      bundleTypes: { js: "js" },
      // TSLint config
      tslint: require("./config-tslint.js")
    };
  },
  commands() {
    return {
      tsLint: {
        //eslint-disable-next-line no-unused-vars
        command: function(crafty, input, cli) {
          // Remove command argument
          process.argv.splice(process.argv.indexOf("tsLint"), 1);

          global.craftyConfig = crafty.config;
          require("./commands/tsLint");
        },
        description: "Lint TypeScript for errors"
      }
    };
  },
  bundleCreator(crafty) {
    const configurators = { js: {} };

    if (
      crafty.config.loadedPresets.some(
        preset => preset.presetName === "@swissquote/crafty-runner-gulp"
      )
    ) {
      configurators.js["gulp/typescript"] = (
        craftyAgain,
        bundle,
        gulp,
        StreamHandler
      ) => {
        gulp.task(
          bundle.taskName,
          createTask(craftyAgain, bundle, StreamHandler)
        );
        craftyAgain.watcher.add(bundle.watch || bundle.source, bundle.taskName);
      };
    }

    return configurators;
  },
  rollup(crafty, bundle, rollupConfig) {
    rollupConfig.input.plugins.typescript = {
      plugin: require("rollup-plugin-typescript2"),
      weight: 20
    };

    // Linting doesn't work well currently in rollup
    // - Errors are imprecise, just outputs "Warnings or errors were found"
    // - If the code can't be parsed, you lose the information of where it failed
    //rollupConfig.input.plugins.tslint = {
    //  plugin: require("rollup-plugin-tslint"),
    //  weight: 0,
    //  options: {
    //    exclude: ["node_modules/**"],
    //    include: ["**/*.ts", "**/*.tsx"],
    //    configuration: createTempFile(JSON.stringify(crafty.config.tslint)),
    //    throwError: true // TODO :: throw only on warning
    //  }
    //};
  },
  jest(crafty, options) {
    options.moduleDirectories.push(MODULES);
    options.transform["^.+\\.tsx?$"] = require.resolve(
      "ts-jest/preprocessor.js"
    );

    options.moduleFileExtensions.push("ts");
    options.moduleFileExtensions.push("tsx");
  },
  webpack(crafty, bundle, chain) {
    chain.resolve.extensions.add(".ts").add(".tsx");
    chain.resolve.modules.add(MODULES);
    chain.resolveLoader.modules.add(MODULES);

    const filename = createTempFile(JSON.stringify(crafty.config.tslint));

    // TypeScript linting
    chain.module
      .rule("lint-ts")
      .pre() // It's important to do this before TypeScript processes the JS.
      .test(/\.tsx?$/)
      .include.add(resolve("js"))
      .end()
      .use("tslint")
      .loader("tslint-loader")
      .options({
        configFile: filename,
        formatter: "stylish"
      });

    // TypeScript
    const tsRule = chain.module.rule("ts");
    tsRule.test(/\.tsx?$/);
    tsRule.exclude.add(/(node_modules|bower_components)/);

    const babelConfigurator = require("@swissquote/babel-preset-swissquote/configurator");
    const babelOptions = babelConfigurator(
      crafty,
      crafty.getEnvironment() === "production" ? "production" : "development",
      bundle,
      {
        deduplicateHelpers: true,
        useESModules: true
      }
    );

    // Cache can be disabled for experimentation and when running Crafty's tests
    if (
      crafty.getEnvironment() === "production" &&
      !process.argv.some(arg => arg === "--no-cache") &&
      !process.env.TESTING_CRAFTY
    ) {
      babelOptions.cacheDirectory = true;
    }

    // EcmaScript 2015+
    tsRule
      .use("babel")
      .loader("babel-loader")
      .options(babelOptions);

    const tsOptions = {
      compilerOptions: {
        // Transpile to esnext so that Babel can apply all its magic
        target: "ESNext",
        // Preserve JSX so babel can optimize it, or add development/debug information
        jsx: "Preserve"
      }
    };

    tsRule
      .use("ts-loader")
      .loader("ts-loader")
      .options(tsOptions);
  }
};
