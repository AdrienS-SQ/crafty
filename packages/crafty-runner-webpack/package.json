{
  "name": "@swissquote/crafty-runner-webpack",
  "version": "1.3.0",
  "license": "Apache-2.0",
  "repository": "https://github.com/swissquote/crafty.git",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "main": "src/index.js",
  "dependencies": {
    "case-sensitive-paths-webpack-plugin": "2.1.2",
    "chalk": "^2.4.1",
    "debug": "^4.1.0",
    "glob-to-regexp": "0.4.0",
    "hash-index": "3.0.0",
    "is-glob": "4.0.0",
    "mkdirp": "0.5.1",
    "pnp-webpack-plugin": "^1.2.0",
    "terser-webpack-plugin": "1.1.0",
    "webpack": "4.24",
    "webpack-chain": "5.0.1",
    "webpack-dev-server": "3.1.10",
    "webpack-merge": "4.1.4",
    "webpack-sources": "1.3.0",
    "webpack-stylish": "0.1.8"
  },
  "engines": {
    "node": ">=8.0.0"
  },
  "devDependencies": {
    "jest": "23.6.0",
    "jest-pnp-resolver": "^1.0.1",
    "jest-resolve": "23.6.0",
    "service-names-port-numbers": "^2.0.0"
  }
}
