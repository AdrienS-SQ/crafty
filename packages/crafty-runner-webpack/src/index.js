const debug = require("debug")("webpack-runner");
const generateTask = require("./webpack_runtime");

module.exports = {
  defaultConfig() {
    return {
      // List of provided libraries, these will not be embedded inside the content
      externals: [],

      // Pass options to uglifyJS, used by both webpack and gulp-uglify
      uglifyJS: { compress: true, sourceMap: true }
    };
  },
  config(config) {
    // Add missing informations to JavaScript configurations
    for (let j in config.js) {
      if (!config.js.hasOwnProperty(j)) {
        continue;
      }

      if (!config.js[j].externals) {
        config.js[j].externals = [];
      }

      // Merge global externals into packages
      if (config.externals && config.externals.length) {
        config.externals.forEach(item => {
          if (config.js[j].externals.indexOf(item) === -1) {
            config.js[j].externals.push(item);
          }
        });
      }
    }

    return config;
  },
  bundleCreator(crafty) {
    return {
      js: {
        webpack: generateTask
      }
    };
  }
};
