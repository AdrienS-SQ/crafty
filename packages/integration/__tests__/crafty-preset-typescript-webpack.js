/* global describe, it, expect, jest */

const fs = require("fs");
const path = require("path");

const rimraf = require("rimraf");
const configuration = require("@swissquote/crafty/src/configuration");
const getCommands = require("@swissquote/crafty/src/commands/index");

const testUtils = require("../utils");

const getCrafty = configuration.getCrafty;

// node-forge 0.6.33 doesn't work with jest.
// but selfsigned is fixed on this version
// as we don't use it for now, we can simply mock it
// https://github.com/jfromaniello/selfsigned/issues/16
jest.mock("node-forge");

it("Loads crafty-preset-typescript and does not register webpack tasks", () => {
  const crafty = getCrafty(["@swissquote/crafty-preset-typescript"], {});

  const loadedPresets = [
    require("@swissquote/crafty-preset-eslint"),
    require("@swissquote/crafty-preset-typescript"),
    { presetName: "crafty.config.js" }
  ];

  expect(crafty.config.loadedPresets).toEqual(loadedPresets);

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([]);
});

it("Loads crafty-preset-typescript, crafty-runner-webpack and registers webpack task", () => {
  const config = { js: { myBundle: { source: "css/style.scss" } } };
  const crafty = getCrafty(
    [
      "@swissquote/crafty-preset-typescript",
      "@swissquote/crafty-runner-webpack"
    ],
    config
  );

  const loadedPresets = [
    require("@swissquote/crafty-preset-eslint"),
    require("@swissquote/crafty-preset-typescript"),
    require("@swissquote/crafty-runner-webpack"),
    Object.assign({ presetName: "crafty.config.js" }, config)
  ];

  expect(crafty.config.loadedPresets).toEqual(loadedPresets);

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([
    "js_myBundle",
    "js",
    "default"
  ]);
});

it("Compiles TypeScript", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-typescript-webpack/compiles"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();
  expect(fs.existsSync("dist/js/1.myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/1.myBundle.min.js.map")).toBeTruthy();

  expect(
    fs.readFileSync("dist/js/myBundle.min.js").toString("utf8")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/1.myBundle.min.js").toString("utf8")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/js/SomeLibrary.d.ts").toString("utf8")
  ).toMatchSnapshot();
});

it("Compiles TypeScript - fork checker", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-typescript-webpack/compiles-forked"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();
  expect(fs.existsSync("dist/js/1.myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/1.myBundle.min.js.map")).toBeTruthy();

  expect(
    fs.readFileSync("dist/js/myBundle.min.js").toString("utf8")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/1.myBundle.min.js").toString("utf8")
  ).toMatchSnapshot();
});

it("Lints TypeScript with webpack", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-typescript-webpack/lints")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Fails gracefully on broken markup", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-typescript-webpack/fails")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myTSBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myTSBundle.min.js.map")).toBeFalsy();
});

it("Fails gracefully on invalid TS", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-typescript-webpack/invalid")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed types
  expect(fs.existsSync("dist/js/myTSBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myTSBundle.min.js.map")).toBeFalsy();
});

it("Fails gracefully on invalid TS - fork checker", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-typescript-webpack/invalid-forked"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed types
  // TODO :: see if it can be done with fork TS Checker
  //expect(fs.existsSync("dist/js/myTSBundle.min.js")).toBeFalsy();
  //expect(fs.existsSync("dist/js/myTSBundle.min.js.map")).toBeFalsy();
});


it("Removes unused classes", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-typescript-webpack/tree-shaking")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();

  const content = fs.readFileSync("dist/js/myBundle.min.js").toString("utf8");

  expect(content.indexOf("From class A") > -1).toBeTruthy();
  expect(content.indexOf("From class B") > -1).toBeFalsy();
});