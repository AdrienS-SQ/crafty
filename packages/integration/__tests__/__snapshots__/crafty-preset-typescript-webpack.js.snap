// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiles TypeScript 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/compiles/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
                Asset       Size  Chunks             Chunk Names
    0.myBundle.min.js  334 bytes       0  [emitted]
      myBundle.min.js    1.81 kB       1  [emitted]  default
0.myBundle.min.js.map     1.7 kB       0  [emitted]
  myBundle.min.js.map    16.7 kB       1  [emitted]  default
    js/Component.d.ts   60 bytes          [emitted]
       js/script.d.ts   73 bytes          [emitted]
  js/SomeLibrary.d.ts  142 bytes          [emitted]
   [0] multi ./js/script.ts 28 bytes {1} [built]
   [1] ./js/script.ts 593 bytes {1} [built]
   [2] ./js/Component.ts 49 bytes {1} [built]
   [3] ./js/SomeLibrary.ts 386 bytes {0} [built]


Compiled with warnings.

myBundle.min.js from UglifyJs
Dropping unused variable result [webpack/bootstrap UNIQUE_ID]

[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles TypeScript 2`] = `
"define(function(){return function(n){function e(t){if(o[t])return o[t].exports;var r=o[t]={i:t,l:!1,exports:{}};return n[t].call(r.exports,r,r.exports,e),r.l=!0,r.exports}var t=window.webpackJsonp;window.webpackJsonp=function(e,o,u){for(var c,i,a=0,f=[];a<e.length;a++)i=e[a],r[i]&&f.push(r[i][0]),r[i]=0;for(c in o)Object.prototype.hasOwnProperty.call(o,c)&&(n[c]=o[c]);for(t&&t(e,o,u);f.length;)f.shift()()};var o={},r={1:0};return e.e=function(n){function t(){i.onerror=i.onload=null,clearTimeout(a);var e=r[n];0!==e&&(e&&e[1](new Error(\\"Loading chunk \\"+n+\\" failed.\\")),r[n]=void 0)}var o=r[n];if(0===o)return new Promise(function(n){n()});if(o)return o[2];var u=new Promise(function(e,t){o=r[n]=[e,t]});o[2]=u;var c=document.getElementsByTagName(\\"head\\")[0],i=document.createElement(\\"script\\");i.type=\\"text/javascript\\",i.charset=\\"utf-8\\",i.async=!0,i.timeout=12e4,e.nc&&i.setAttribute(\\"nonce\\",e.nc),i.src=e.p+\\"\\"+n+\\".myBundle.min.js\\";var a=setTimeout(t,12e4);return i.onerror=i.onload=t,c.appendChild(i),u},e.m=n,e.c=o,e.d=function(n,t,o){e.o(n,t)||Object.defineProperty(n,t,{configurable:!1,enumerable:!0,get:o})},e.n=function(n){var t=n&&n.__esModule?function(){return n.default}:function(){return n};return e.d(t,\\"a\\",t),t},e.o=function(n,e){return Object.prototype.hasOwnProperty.call(n,e)},e.p=\\"\\",e.oe=function(n){throw console.error(n),n},e(e.s=0)}([function(n,e,t){n.exports=t(1)},function(n,e,t){\\"use strict\\";Object.defineProperty(e,\\"__esModule\\",{value:!0});var o=t(2),r=function(){function n(){console.log(\\"init class\\")}return n.prototype.method=function(){console.log(Object(o.a)(2,4)),t.e(0).then(t.bind(null,3)).then(function(n){var e=new n.default;console.log(e.add(2,4))})},n}();e.default=r},function(n,e,t){\\"use strict\\";function o(n,e){return n+e}e.a=o}])});
//# sourceMappingURL=myBundle.min.js.map"
`;

exports[`Compiles TypeScript 3`] = `
"webpackJsonp([0],{3:function(t,e,n){\\"use strict\\";Object.defineProperty(e,\\"__esModule\\",{value:!0});var r=function(){function t(){}return t.prototype.add=function(t,e){return t+e},t.prototype.substract=function(t,e){return t-e},t.prototype.divide=function(t,e){return t/e},t}();e.default=r}});
//# sourceMappingURL=0.myBundle.min.js.map"
`;

exports[`Fails gracefully on broken markup 1`] = `
Object {
  "status": 1,
  "stderr": "[js_myTSBundle] (!) __PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/fails/js/typescript.ts(14,1): syntax error TS1068 Unexpected token. A constructor, method, accessor, or property was expected.
js/typescript.ts
[__:__:__] 'js_myTSBundle' errored after ____ ms
[__:__:__] Error: __PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/fails/js/typescript.ts(14,1): syntax error TS1068 Unexpected token. A constructor, method, accessor, or property was expected.
    ...stacktrace...
[__:__:__] 'js' errored after ____ ms
[__:__:__] 'default' errored after ____ ms
",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/fails/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myTSBundle' ...
",
}
`;

exports[`Lints TypeScript with webpack 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/lints/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
                Asset       Size  Chunks             Chunk Names
    0.myBundle.min.js  334 bytes       0  [emitted]
      myBundle.min.js    1.81 kB       1  [emitted]  default
0.myBundle.min.js.map     1.7 kB       0  [emitted]
  myBundle.min.js.map    16.4 kB       1  [emitted]  default
    js/Component.d.ts   60 bytes          [emitted]
       js/script.d.ts   73 bytes          [emitted]
  js/SomeLibrary.d.ts  142 bytes          [emitted]
   [0] multi ./js/script.ts 28 bytes {1} [built]
   [1] ./js/script.ts 448 bytes {1} [built] [1 warning]
   [2] ./js/Component.ts 49 bytes {1} [built] [1 warning]
   [3] ./js/SomeLibrary.ts 386 bytes {0} [built] [1 warning]


Compiled with warnings.

./js/Component.ts
__PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/lints/js/Component.ts:2:24
ERROR: 2:24  whitespace       missing whitespace
ERROR: 3:18  no-multi-spaces  Multiple spaces found before 'b'.

 @ ./js/script.ts 1:0-35
 @ multi ./js/script.ts

./js/SomeLibrary.ts
__PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/lints/js/SomeLibrary.ts:12:5
ERROR: 12:5  one-line  misplaced opening brace

 @ ./js/script.ts 8:8-31
 @ multi ./js/script.ts

./js/script.ts
__PATH__/packages/integration/fixtures/crafty-preset-typescript-webpack/lints/js/script.ts:6:5
ERROR: 6:5   no-console  Calls to 'console.log' are not allowed.
ERROR: 10:5  no-console  Calls to 'console.log' are not allowed.
ERROR: 13:5  one-line    misplaced opening brace
ERROR: 16:7  no-console  Calls to 'console.log' are not allowed.
ERROR: 17:7  semicolon   Missing semicolon

 @ multi ./js/script.ts

myBundle.min.js from UglifyJs
Dropping unused variable result [webpack/bootstrap UNIQUE_ID]

[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;
